src/groups/mqb/mqbstat/mqbstat_statcontroller.cpp

```diff
@@ -233,26 +241,51 @@ void StatController::initializeStats()
             false)));
 }
 
-void StatController::captureStats(mqbcmd::StatResult* result)
+void StatController::captureStatsAndSemaphorePost(
+    mqbcmd::StatResult*                  result,
+    bslmt::Semaphore*                    semaphore,
+    const mqbcmd::EncodingFormat::Value& encoding)
 {
-    // This must execute in the *SCHEDULER* thread.
+    // executed by the *SCHEDULER* thread
 
     if (d_allocatorsStatContext_p) {
         // When using test allocator, we don't have a stat context
         d_allocatorsStatContext_p->snapshot();
     }
 
-    mwcu::MemOutStream os;
-    d_printer_mp->printStats(os);
-    result->makeStats(os.str());
-}
+    switch (encoding) {
+    case mqbcmd::EncodingFormat::TEXT: {
+        mwcu::MemOutStream os;
+        d_printer_mp->printStats(os);
+        result->makeStats() = os.str();
+    } break;  // BREAK
+
+    case mqbcmd::EncodingFormat::JSON_COMPACT: BSLS_ANNOTATION_FALLTHROUGH;
+    case mqbcmd::EncodingFormat::JSON_PRETTY: {
+        // Make an unscheduled snapshot, but do not notify stats consumers
+        // since it's not necessary
+        const bool savedNextSnapshot = snapshot();
```

Comment:
Why is it necessary to make a snapshot before printing as JSON?

Reply:
We specifically want to use this code path in integration tests, to check metrics generation. So we want to get the latest stats. To do this, we either `sleep` in IT with hope that it's enough time to make a new snapshot, or add a sync point like I did.
Note that this code path is for stats requested by JSON only, so the previous cmd `STAT SHOW` will not be affected.

Reply:
Extended the comment